src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,353 @@
+//! Memory allocation APIs
+
+#![stable(feature = "alloc_module", since = "1.28.0")]
+
+mod global;
+mod layout;
+
+#[stable(feature = "global_alloc", since = "1.28.0")]
+pub use self::global::GlobalAlloc;
+#[stable(feature = "alloc_layout", since = "1.28.0")]
+pub use self::layout::{Layout, LayoutErr};
+
+use crate::fmt;
+use crate::ptr::{self, NonNull};
+
+/// The `AllocErr` error indicates an allocation failure
+/// that may be due to resource exhaustion or to
+/// something wrong when combining the given input arguments with this
+/// allocator.
+#[unstable(feature = "allocator_api", issue = "32838")]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AllocErr;
+
+// (we need this for downstream impl of trait Error)
+#[unstable(feature = "allocator_api", issue = "32838")]
+impl fmt::Display for AllocErr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("memory allocation failed")
+    }
+}
+
+/// A desired initial state for allocated memory.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub enum AllocInit {
+    /// The contents of the new memory are undefined.
+    ///
+    /// Reading uninitialized memory is Undefined Behavior; it must be initialized before use.
+    Uninitialized,
+    /// The new memory is guaranteed to be zeroed.
+    Zeroed,
+}
+
+/// Represents a block of allocated memory returned by an allocator.
+#[derive(Debug, Copy, Clone)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub struct MemoryBlock {
+    pub ptr: NonNull<u8>,
+    pub size: usize,
+}
+
+impl MemoryBlock {
+    /// Initialize the memory block like specified by `init`.
+    ///
+    /// This behaves like calling [`MemoryBlock::initialize_offset(ptr, layout, 0)`][off].
+    ///
+    /// [off]: MemoryBlock::init_offset
+    ///
+    /// [*fit*]: trait.AllocRef.html#memory-fitting
+    #[inline]
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    pub fn init(&mut self, init: AllocInit) {
```

Comment:
Having these methods on `MemoryBlock` feels strange since they will only ever be called by implementers of `AllocRef`, not users. I think moving them to methods on `AllocInit` would be more suitable.